cmd/rpcdaemon/commands/otterscan_api.go

```diff
@@ -112,6 +114,164 @@ func (api *OtterscanAPIImpl) getTransactionByHash(ctx context.Context, tx kv.Tx,
+		return vm.STATICCALL, nil
+	case "DELEGATECALL":
+		return vm.DELEGATECALL, nil
+	case "CALLCODE":
+		return vm.CALLCODE, nil
+	case "CREATE":
+		return vm.CREATE, nil
+	case "CREATE2":
+		return vm.CREATE2, nil
+	case "SELFDESTRUCT":
+		return vm.SELFDESTRUCT, nil
+	}
+	return vm.INVALID, fmt.Errorf("unable to translate %s", typStr)
+}
+
+func (api *OtterscanAPIImpl) translateCaptureEnter(gethTrace *GethTrace, tracer vm.EVMLogger, vmenv *vm.EVM) error {
+	from := common.HexToAddress(gethTrace.From)
+	to := common.HexToAddress(gethTrace.To)
+	input, err := hexutil.Decode(gethTrace.Input)
+	if err != nil {
+		if err != hexutil.ErrEmptyString {
+			return err
+		}
+		input = []byte{}
+	}
+	valueBig, err := hexutil.DecodeBig(gethTrace.Value)
+	if err != nil {
+		if err != hexutil.ErrEmptyString {
+			return err
+		}
+		valueBig = big.NewInt(0)
+	}
+	value, _ := uint256.FromBig(valueBig)
+	gas, err := hexutil.DecodeUint64(gethTrace.Gas)
+	if err != nil {
+		return err
+	}
+	typStr := gethTrace.Type
+	typ, err := api.translateOpcode(typStr)
+	if err != nil {
+		return err
+	}
+	_, isPrecompile := vmenv.Precompile(to)
+	tracer.CaptureEnter(typ, from, to, isPrecompile, false, input, gas, value, nil)
+	return nil
+}
+
+func (api *OtterscanAPIImpl) translateCaptureExit(gethTrace *GethTrace, tracer vm.EVMLogger) error {
+	usedGas, err := hexutil.DecodeUint64(gethTrace.GasUsed)
+	if err != nil {
+		return err
+	}
+	output, err := hexutil.Decode(gethTrace.Output)
+	if err != nil {
+		if err != hexutil.ErrEmptyString {
+			return err
+		}
+		output = []byte{}
+	}
+	err = errors.New(gethTrace.Error)
+	tracer.CaptureExit(output, usedGas, err)
+	return nil
+}
+
+func (api *OtterscanAPIImpl) translateRelayTraceResult(gethTrace *GethTrace, tracer vm.EVMLogger, chainConfig *chain.Config) error {
+	vmenv := vm.NewEVM(evmtypes.BlockContext{}, evmtypes.TxContext{}, nil, chainConfig, vm.Config{})
+	type traceWithIndex struct {
+		gethTrace *GethTrace
+		idx       int // children index
+	}
+	callStacks := make([]*traceWithIndex, 0)
+	started := false
+	rootIndex := 0
+	var trace *GethTrace = gethTrace
+	// iterative postorder traversal
+	for trace != nil || len(callStacks) > 0 {
+		if trace != nil {
+			// push back
+			callStacks = append(callStacks, &traceWithIndex{trace, rootIndex})
+			if !started {
+				started = true
+				if err := api.translateCaptureStart(trace, tracer, vmenv); err != nil {
+					return err
+				}
+			} else {
+				if err := api.translateCaptureEnter(trace, tracer, vmenv); err != nil {
+					return err
+				}
+			}
+			rootIndex = 0
+			if len(trace.Calls) > 0 {
+				trace = trace.Calls[0]
+			} else {
+				trace = nil
+			}
+			continue
+		}
+		// pop back
+		top := callStacks[len(callStacks)-1]
+		callStacks = callStacks[:len(callStacks)-1]
+		if err := api.translateCaptureExit(top.gethTrace, tracer); err != nil {
+			return err
+		}
+		for len(callStacks) > 0 && top.idx == len(callStacks[len(callStacks)-1].gethTrace.Calls)-1 {
```

Comment:
It's a little bit confusing to understand the meaning and purpose of `idx`. Please add some comments to explain the defenition of `idx`, managing `rootindex`, and the condition on L259.